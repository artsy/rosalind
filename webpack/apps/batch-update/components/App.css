#root {
  position: absolute;
  left: 0;
  top: 50px;
  right: 0;
  bottom: 0;
}

/* basic layout */

.App {
  min-height: calc(100vh - 50px);
  display: flex;
  flex-flow: row wrap;
  align-content: flex-start;
  align-items: stretch;
}

.SearchForm, .SearchResults {
  padding: 1em;
}

.SearchForm {
  flex-grow: 1;
  flex-basis: 150px;
  border-right: solid 1px #eee;
  margin-top: ;
}

.SearchResults {
  position: relative;
  flex-grow: 5;
  flex-basis: 400px;
}

/* components */

.SearchForm-SectionHeader {
  color: #999;
  text-transform: uppercase;
  letter-spacing: 0.1em;
  font-size: 0.75em;
  font-family:'ITC Avant Garde Gothic W01';
  font-weight: 400;
  margin-top: 1.5em;
}

.SelectedGene,
.SelectedTag,
.SelectedPartner, 
.SelectedFair {
  font-weight: bold;
  margin: 0.25em 0;
}

.SelectedGene-remove,
.SelectedTag-remove,
.SelectedPartner-remove, 
.SelectedFair-remove {
  display: inline-block;
  height: 2em;
  width: 2em;
  line-height: 2em;
  text-align: center;
  color: #ddd;
  margin-left: 0.5em;
  border-radius: 1em;
  text-decoration: none;
}

.SelectedGene-remove:hover,
.SelectedTag-remove:hover,
.SelectedPartner-remove:hover, 
.SelectedFair-remove:hover {
  color: #333;
  cursor: pointer;
  background: #eee;
}

.FilterOptions {
  font-size: 80%;
  color: #999;
  line-height: 140%;
  margin-top: 2em;
  opacity: 0.25;
  transition: opacity 0.75s;
}

.FilterOptions:hover {
  opacity: 1;
  transition: opacity 0.25s;
}

.PublishedFilter,
.DeletedFilter,
.GenomedFilter {
  margin-top: 1em;
}

.PublishedFilter a,
.DeletedFilter a,
.GenomedFilter a {
  color: #ccc;
  margin-right: 0.5em;
}

.PublishedFilter a.active,
.DeletedFilter a.active,
.GenomedFilter a.active {
  color: #333;
  font-weight: bold;
  text-decoration: none;
  cursor: default;;
}

.SearchResults {
  display: flex;
  flex-flow: row wrap;
  align-content: flex-start;
}

.ArtworkResult {
  /*outline: solid 1px red;*/
  min-width: 115px;
  max-width: calc(20% - 1.5em);
  margin: 0.75em;
  flex: 1 1 calc(20% - 1.5em);
}

.ArtworkResult img {
  width: 100%;
}

.ArtworkResult figcaption {
  position: relative;
  font-size: 80%;
  max-height: 3.75em;
  line-height: 1.25em;
  overflow: hidden;
}

.ArtworkPreview-Overlay {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: hsla(0, 0%, 0%, 0.5);
  z-index: 1;

  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
}

.ArtworkPreview {
  /*flex: 0 0 66%;*/
  box-sizing: border-box;
  padding: 2em;
  max-height: 90%;
  z-index: 2;
  background: white;
  box-shadow: 0 0 20px hsla(0, 0%, 0%, 0.5);

  display: flex;
  flex-direction: row;
  justify-content: space-between;
}

.ArtworkPreview-Image {
  flex: 0 1 40%;
}

.ArtworkPreview-Image img {
  /*min-width: 500px;*/
}

.ArtworkPreview-Details {
  max-width: 20em;
  margin-left: 1em;
  flex: 0 1 60%;
}

.ArtworkPreview-Name {
  font-weight: bold;
}

.ArtworkPreview-Status {
  font-size: 0.9em;
  line-height: 125%;
  margin: 1em 0;
}

.ArtworkPreview-Links a {
  margin: 0 0.25em;
}
