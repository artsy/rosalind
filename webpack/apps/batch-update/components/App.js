import React from 'react'
import SearchForm from './SearchForm'
import SearchResults from './SearchResults'
import { buildElasticsearchQuery } from '../helpers/elasticsearch'
import { matchArtworks } from 'lib/rosalind-api'
import { Wrapper, Sidebar, Content } from './Layout'

const findByName = (items, item) => items.find(i => i.name === item.name)

class App extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      genes: [],
      tags: [],
      partner: null,
      fair: null,
      publishedFilter: 'SHOW_ALL',
      deletedFilter: 'SHOW_ALL',
      genomedFilter: 'SHOW_ALL',
      artworks: [],
      previewedArtwork: null,
      isLoading: false,
      totalHits: null,
      size: 100
    }
    this.onRemoveGene = this.onRemoveGene.bind(this)
    this.onAddGene = this.onAddGene.bind(this)
    this.onRemoveTag = this.onRemoveTag.bind(this)
    this.onAddTag = this.onAddTag.bind(this)
    this.onSetPartner = this.onSetPartner.bind(this)
    this.onClearPartner = this.onClearPartner.bind(this)
    this.onSetFair = this.onSetFair.bind(this)
    this.onClearFair = this.onClearFair.bind(this)

    this.onSetPublishedFilter = this.onSetPublishedFilter.bind(this)
    this.onSetDeletedFilter = this.onSetDeletedFilter.bind(this)
    this.onSetGenomedFilter = this.onSetGenomedFilter.bind(this)

    this.onPreviewArtwork = this.onPreviewArtwork.bind(this)
    this.onPreviewPrevious = this.onPreviewPrevious.bind(this)
    this.onPreviewNext = this.onPreviewNext.bind(this)

    this.fetchArtworks = this.fetchArtworks.bind(this)
    this.fetchMoreArtworks = this.fetchMoreArtworks.bind(this)
  }

  componentWillMount () {
    if (this.state.genes.length || this.state.tags.length) {
      this.fetchArtworks()
    }
  }

  componentDidUpdate (prevProps, prevState) {
    if (
      (this.state.genes !== prevState.genes) ||
      (this.state.tags !== prevState.tags) ||
      (this.state.partner !== prevState.partner) ||
      (this.state.fair !== prevState.fair) ||
      (this.state.publishedFilter !== prevState.publishedFilter) ||
      (this.state.deletedFilter !== prevState.deletedFilter) ||
      (this.state.genomedFilter !== prevState.genomedFilter)
     ) {
      this.fetchArtworks()
    }
  }

  fetchArtworks () {
    const { genes, tags, partner, fair } = this.state
    if ((genes.length === 0) && (tags.length === 0) && (partner === null) && (fair === null)) {
      this.setState({ artworks: [], totalHits: 0 })
    } else {
      const { publishedFilter, deletedFilter, genomedFilter, size } = this.state
      const query = buildElasticsearchQuery({ genes, tags, partner, fair, publishedFilter, deletedFilter, genomedFilter, size })
      this.setState({ isLoading: true })
      matchArtworks(query).then(hits => {
        const totalHits = hits.total
        const artworks = hits.hits.map(hit => hit._source)
        this.setState({ artworks, totalHits, isLoading: false })
      })
    }
  }

  fetchMoreArtworks () {
    const { genes, tags, partner, fair, publishedFilter, deletedFilter, genomedFilter } = this.state
    const { artworks, size } = this.state
    const from = artworks.length
    const query = buildElasticsearchQuery({ genes, tags, partner, fair, publishedFilter, deletedFilter, genomedFilter, from, size })
    matchArtworks(query).then(hits => {
      const totalHits = hits.total
      const moreArtworks = hits.hits.map(hit => hit._source)
      this.setState({ artworks: [...artworks, ...moreArtworks], totalHits })
    })
  }

  onRemoveGene (geneName) {
    const { genes } = this.state
    this.setState({
      genes: genes.filter(g => g.name !== geneName)
    })
  }

  onAddGene (gene) {
    const { genes } = this.state
    findByName(genes, gene) || this.setState({
      genes: genes.concat(gene)
    })
  }

  onRemoveTag (tagName) {
    const { tags } = this.state
    this.setState({
      tags: tags.filter(t => t.name !== tagName)
    })
  }

  onAddTag (tag) {
    const { tags } = this.state
    findByName(tags, tag) || this.setState({
      tags: tags.concat(tag)
    })
  }

  onSetPartner (partner) {
    this.setState({ partner })
  }

  onClearPartner () {
    this.setState({ partner: null })
  }

  onSetFair (fair) {
    this.setState({ fair })
  }

  onClearFair () {
    this.setState({ fair: null })
  }

  onSetPublishedFilter (filterValue) {
    this.setState({ publishedFilter: filterValue })
  }

  onSetDeletedFilter (filterValue) {
    this.setState({ deletedFilter: filterValue })
  }

  onSetGenomedFilter (filterValue) {
    this.setState({ genomedFilter: filterValue })
  }

  onPreviewArtwork (artwork) {
    this.setState({ previewedArtwork: artwork })
  }

  onPreviewPrevious () {
    const curr = this.state.artworks.indexOf(this.state.previewedArtwork)
    const prev = Math.max(0, curr - 1)
    const artwork = this.state.artworks[prev]
    this.setState({ previewedArtwork: artwork })
  }

  onPreviewNext () {
    const curr = this.state.artworks.indexOf(this.state.previewedArtwork)
    const next = Math.min(this.state.artworks.length, curr + 1)
    const artwork = this.state.artworks[next]
    this.setState({ previewedArtwork: artwork })
  }

  render () {
    const { genes, tags, partner, fair, artworks, totalHits, previewedArtwork, isLoading } = this.state
    return (
      <Wrapper>
        <Sidebar>
          <SearchForm
            genes={genes}
            tags={tags}
            partner={partner}
            fair={fair}
            onRemoveGene={this.onRemoveGene}
            onAddGene={this.onAddGene}
            onRemoveTag={this.onRemoveTag}
            onAddTag={this.onAddTag}
            onSetPartner={this.onSetPartner}
            onClearPartner={this.onClearPartner}
            onSetFair={this.onSetFair}
            onClearFair={this.onClearFair}
            publishedFilter={this.state.publishedFilter}
            onSetPublishedFilter={this.onSetPublishedFilter}
            deletedFilter={this.state.deletedFilter}
            onSetDeletedFilter={this.onSetDeletedFilter}
            genomedFilter={this.state.genomedFilter}
            onSetGenomedFilter={this.onSetGenomedFilter}
          />
        </Sidebar>

        <Content>
          <SearchResults
            artworks={artworks}
            previewedArtwork={previewedArtwork}
            isLoading={isLoading}
            totalHits={totalHits}
            onPreviewArtwork={this.onPreviewArtwork}
            onPreviewPrevious={this.onPreviewPrevious}
            onPreviewNext={this.onPreviewNext}
            onLoadMore={this.fetchMoreArtworks}
          />
        </Content>
      </Wrapper>
    )
  }
}

export default App
