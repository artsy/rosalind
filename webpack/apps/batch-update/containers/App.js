import React from 'react'
import SearchForm from './SearchForm'
import SearchResults from './SearchResults'
import { buildElasticsearchQuery } from '../helpers/elasticsearch'
import { matchArtworks } from 'lib/rosalind-api'
import { Wrapper, Sidebar, Content } from './Layout'
import FullScreenModal from './FullScreenModal'

const findByName = (items, item) => items.find(i => i.name === item.name)

class App extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      artworks: [],
      createdAfterDate: null,
      createdBeforeDate: null,
      deletedFilter: 'SHOW_ALL',
      fair: null,
      genes: [],
      genomedFilter: 'SHOW_ALL',
      isLoading: false,
      isSpecifyingBatchUpdate: false,
      partner: null,
      previewedArtwork: null,
      publishedFilter: 'SHOW_ALL',
      selectedArtworkIds: [],
      size: 100,
      tags: [],
      totalHits: null
    }

    this.shouldComponentUpdate = this.shouldComponentUpdate.bind(this)
    this.onRemoveGene = this.onRemoveGene.bind(this)
    this.onAddGene = this.onAddGene.bind(this)
    this.onRemoveTag = this.onRemoveTag.bind(this)
    this.onAddTag = this.onAddTag.bind(this)
    this.onSetPartner = this.onSetPartner.bind(this)
    this.onClearPartner = this.onClearPartner.bind(this)
    this.onSetFair = this.onSetFair.bind(this)
    this.onClearFair = this.onClearFair.bind(this)
    this.onAddCreatedAfterDate = this.onAddCreatedAfterDate.bind(this)
    this.onClearCreatedAfterDate = this.onClearCreatedAfterDate.bind(this)
    this.onAddCreatedBeforeDate = this.onAddCreatedBeforeDate.bind(this)
    this.onClearCreatedBeforeDate = this.onClearCreatedBeforeDate.bind(this)

    this.onSetPublishedFilter = this.onSetPublishedFilter.bind(this)
    this.onSetDeletedFilter = this.onSetDeletedFilter.bind(this)
    this.onSetGenomedFilter = this.onSetGenomedFilter.bind(this)

    this.onToggleArtwork = this.onToggleArtwork.bind(this)
    this.onSelectAllArtworks = this.onSelectAllArtworks.bind(this)
    this.onDeselectAllArtworks = this.onDeselectAllArtworks.bind(this)
    this.onPreviewArtwork = this.onPreviewArtwork.bind(this)
    this.onPreviewPrevious = this.onPreviewPrevious.bind(this)
    this.onPreviewNext = this.onPreviewNext.bind(this)

    this.fetchArtworks = this.fetchArtworks.bind(this)
    this.fetchMoreArtworks = this.fetchMoreArtworks.bind(this)

    this.onOpenBatchUpdate = this.onOpenBatchUpdate.bind(this)
    this.onDismissBatchUpdate = this.onDismissBatchUpdate.bind(this)
  }

  componentWillMount () {
    if (this.state.genes.length || this.state.tags.length) {
      this.fetchArtworks()
    }
  }

  componentDidUpdate (_prevProps, prevState) {
    if (this.shouldComponentUpdate(prevState)) {
      this.fetchArtworks()
    }
  }

  shouldComponentUpdate (prevState) {
    return (
      (this.state.createdAfterDate !== prevState.createdAfterDate) ||
      (this.state.createdBeforeDate !== prevState.createdBeforeDate) ||
      (this.state.deletedFilter !== prevState.deletedFilter) ||
      (this.state.fair !== prevState.fair) ||
      (this.state.genes !== prevState.genes) ||
      (this.state.genomedFilter !== prevState.genomedFilter) ||
      (this.state.partner !== prevState.partner) ||
      (this.state.publishedFilter !== prevState.publishedFilter) ||
      (this.state.tags !== prevState.tags)
    )
  }

  fetchArtworks () {
    const {
      createdAfterDate,
      createdBeforeDate,
      genes,
      tags,
      partner,
      fair
    } = this.state

    if ((genes.length === 0) &&
      (tags.length === 0) &&
      (partner === null) && (fair === null)
    ) {
      this.setState({ artworks: [], totalHits: 0 })
    } else {
      const { publishedFilter, deletedFilter, genomedFilter, size } = this.state

      const query = buildElasticsearchQuery({
        createdAfterDate,
        createdBeforeDate,
        deletedFilter,
        fair,
        genes,
        genomedFilter,
        partner,
        publishedFilter,
        size,
        tags
      })

      this.setState({ isLoading: true })
      matchArtworks(query).then(hits => {
        const totalHits = hits.total
        const artworks = hits.hits.map(hit => hit._source)
        this.setState({ artworks, totalHits, isLoading: false })
      })
    }
  }

  fetchMoreArtworks () {
    const {
      createdAfterDate,
      createdBeforeDate,
      deletedFilter,
      fair,
      genes,
      genomedFilter,
      partner,
      publishedFilter,
      tags
    } = this.state

    const { artworks, size } = this.state

    const from = artworks.length

    const query = buildElasticsearchQuery({
      createdAfterDate,
      createdBeforeDate,
      deletedFilter,
      fair,
      from,
      genes,
      genomedFilter,
      partner,
      publishedFilter,
      size,
      tags
    })

    matchArtworks(query).then(hits => {
      const totalHits = hits.total
      const moreArtworks = hits.hits.map(hit => hit._source)
      this.setState({ artworks: [...artworks, ...moreArtworks], totalHits })
    })
  }

  onRemoveGene (geneName) {
    const { genes } = this.state
    this.setState({
      genes: genes.filter(g => g.name !== geneName)
    })
  }

  onAddGene (gene) {
    const { genes } = this.state
    findByName(genes, gene) || this.setState({
      genes: genes.concat(gene)
    })
  }

  onRemoveTag (tagName) {
    const { tags } = this.state
    this.setState({
      tags: tags.filter(t => t.name !== tagName)
    })
  }

  onAddTag (tag) {
    const { tags } = this.state
    findByName(tags, tag) || this.setState({
      tags: tags.concat(tag)
    })
  }

  onSetPartner (partner) {
    this.setState({ partner })
  }

  onClearPartner () {
    this.setState({ partner: null })
  }

  onSetFair (fair) {
    this.setState({ fair })
  }

  onClearFair () {
    this.setState({ fair: null })
  }

  onAddCreatedAfterDate (createdAfterDate) {
    this.setState({ createdAfterDate })
  }

  onClearCreatedAfterDate () {
    this.setState({ createdAfterDate: null })
  }

  onAddCreatedBeforeDate (createdBeforeDate) {
    this.setState({ createdBeforeDate })
  }

  onClearCreatedBeforeDate () {
    this.setState({ createdBeforeDate: null })
  }

  onSetPublishedFilter (filterValue) {
    this.setState({ publishedFilter: filterValue })
  }

  onSetDeletedFilter (filterValue) {
    this.setState({ deletedFilter: filterValue })
  }

  onSetGenomedFilter (filterValue) {
    this.setState({ genomedFilter: filterValue })
  }

  onToggleArtwork (artwork) {
    const { selectedArtworkIds } = this.state
    if (selectedArtworkIds.indexOf(artwork.id) > -1) {
      this.setState({
        selectedArtworkIds: selectedArtworkIds.filter(id => id !== artwork.id)
      })
    } else {
      this.setState({
        selectedArtworkIds: [...selectedArtworkIds, artwork.id]
      })
    }
  }

  onSelectAllArtworks () {
    const { artworks } = this.state
    this.setState({
      selectedArtworkIds: artworks.map(a => a.id)
    })
  }

  onDeselectAllArtworks () {
    this.setState({
      selectedArtworkIds: []
    })
  }

  onPreviewArtwork (artwork) {
    this.setState({ previewedArtwork: artwork })
  }

  onPreviewPrevious () {
    const curr = this.state.artworks.indexOf(this.state.previewedArtwork)
    const prev = Math.max(0, curr - 1)
    const artwork = this.state.artworks[prev]
    this.setState({ previewedArtwork: artwork })
  }

  onPreviewNext () {
    const curr = this.state.artworks.indexOf(this.state.previewedArtwork)
    const next = Math.min(this.state.artworks.length, curr + 1)
    const artwork = this.state.artworks[next]
    this.setState({ previewedArtwork: artwork })
  }

  onOpenBatchUpdate () {
    this.setState({ isSpecifyingBatchUpdate: true })
  }

  onDismissBatchUpdate () {
    this.setState({ isSpecifyingBatchUpdate: false })
  }

  render () {
    const {
      artworks,
      createdAfterDate,
      createdBeforeDate,
      deletedFilter,
      fair,
      genes,
      genomedFilter,
      isLoading,
      isSpecifyingBatchUpdate,
      partner,
      previewedArtwork,
      publishedFilter,
      selectedArtworkIds,
      tags,
      totalHits
    } = this.state

    return (
      <Wrapper>
        <Sidebar>
          <SearchForm
            createdAfterDate={createdAfterDate}
            createdBeforeDate={createdBeforeDate}
            deletedFilter={deletedFilter}
            fair={fair}
            genes={genes}
            genomedFilter={genomedFilter}
            onAddCreatedAfterDate={this.onAddCreatedAfterDate}
            onAddCreatedBeforeDate={this.onAddCreatedBeforeDate}
            onAddGene={this.onAddGene}
            onAddTag={this.onAddTag}
            onClearCreatedAfterDate={this.onClearCreatedAfterDate}
            onClearCreatedBeforeDate={this.onClearCreatedBeforeDate}
            onClearFair={this.onClearFair}
            onClearPartner={this.onClearPartner}
            onOpenBatchUpdate={this.onOpenBatchUpdate}
            onRemoveGene={this.onRemoveGene}
            onRemoveTag={this.onRemoveTag}
            onSetDeletedFilter={this.onSetDeletedFilter}
            onSetFair={this.onSetFair}
            onSetGenomedFilter={this.onSetGenomedFilter}
            onSetPartner={this.onSetPartner}
            onSetPublishedFilter={this.onSetPublishedFilter}
            partner={partner}
            publishedFilter={publishedFilter}
            selectedArtworksCount={selectedArtworkIds.length}
            tags={tags}
          />
        </Sidebar>

        <Content>
          <SearchResults
            artworks={artworks}
            selectedArtworkIds={selectedArtworkIds}
            previewedArtwork={previewedArtwork}
            isLoading={isLoading}
            totalHits={totalHits}
            onToggleArtwork={this.onToggleArtwork}
            onSelectAllArtworks={this.onSelectAllArtworks}
            onDeselectAllArtworks={this.onDeselectAllArtworks}
            onPreviewArtwork={this.onPreviewArtwork}
            onPreviewPrevious={this.onPreviewPrevious}
            onPreviewNext={this.onPreviewNext}
            onLoadMore={this.fetchMoreArtworks}
          />
        </Content>

        <FullScreenModal isOpen={isSpecifyingBatchUpdate} onDismiss={this.onDismissBatchUpdate}>
          <p>{selectedArtworkIds.length} works selected</p>
          <a href='#' onClick={this.onDismissBatchUpdate}>cancel</a>
        </FullScreenModal>
      </Wrapper>
    )
  }
}

export default App
