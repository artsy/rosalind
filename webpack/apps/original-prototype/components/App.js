import React from 'react'
import SearchForm from './SearchForm.js'
import SearchResults from './SearchResults.js'
import { matchArtworks } from 'lib/rosalind-api'
import './App.css'

const findByName = (items, item) => items.find(i => i.name === item.name)

class App extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      genes: [],
      tags: [],
      partner: null,
      fair: null,
      publishedFilter: 'SHOW_ALL',
      deletedFilter: 'SHOW_ALL',
      genomedFilter: 'SHOW_ALL',
      artworks: [],
      previewedArtwork: null
    }
    this.onRemoveGene = this.onRemoveGene.bind(this)
    this.onAddGene = this.onAddGene.bind(this)
    this.onRemoveTag = this.onRemoveTag.bind(this)
    this.onAddTag = this.onAddTag.bind(this)
    this.onSetPartner = this.onSetPartner.bind(this)
    this.onClearPartner = this.onClearPartner.bind(this)
    this.onSetFair = this.onSetFair.bind(this)
    this.onClearFair = this.onClearFair.bind(this)

    this.onSetPublishedFilter = this.onSetPublishedFilter.bind(this)
    this.onSetDeletedFilter = this.onSetDeletedFilter.bind(this)
    this.onSetGenomedFilter = this.onSetGenomedFilter.bind(this)

    this.onPreviewArtwork = this.onPreviewArtwork.bind(this)
    this.onPreviewPrevious = this.onPreviewPrevious.bind(this)
    this.onPreviewNext = this.onPreviewNext.bind(this)
  }

  fetchArtworks () {
    const { genes, tags, partner, fair } = this.state
    if ((genes.length === 0) && (tags.length === 0) && (partner === null) && (fair === null)) {
      this.setState({ artworks: [] })
    } else {
      const query = this.buildElasticSearchQuery()
      matchArtworks(query).then(artworks => {
        this.setState({ artworks: artworks })
      })
    }
  }

  buildElasticSearchQuery () {
    const { genes, tags, partner, fair } = this.state
    const geneMatches = genes.map(g => { return { 'match': { 'genes': g.name } } })
    const tagMatches = tags.map(t => { return { 'match': { 'tags': t.name } } })
    const filterMatches = this.buildFilterMatches()
    const partnerMatch = partner ? { 'match': { 'partner_id': partner.id } } : null
    const fairMatch = fair ? { 'match': { 'fair_ids': fair.id } } : null
    return {
      'query': {
        'bool': {
          'must': [...geneMatches, ...tagMatches, ...filterMatches, partnerMatch, fairMatch].filter(m => m !== null)
        }
      },
      'sort': [
          { 'published_at': 'desc' },
          { 'id': 'desc' }
      ],
      'size': 100
    }
  }

  buildFilterMatches () {
    let matches = [
      this.publishedMatcher(),
      this.deletedMatcher(),
      this.genomedMatcher()
    ]
    return matches
  }

  publishedMatcher () {
    switch (this.state.publishedFilter) {
      case 'SHOW_PUBLISHED':
        return { 'match': { 'published': true } }
      case 'SHOW_NOT_PUBLISHED':
        return { 'match': { 'published': false } }
      default:
        return null
    }
  }

  deletedMatcher () {
    switch (this.state.deletedFilter) {
      case 'SHOW_DELETED':
        return { 'match': { 'deleted': true } }
      case 'SHOW_NOT_DELETED':
        return { 'match': { 'deleted': false } }
      default:
        return null
    }
  }

  genomedMatcher () {
    switch (this.state.genomedFilter) {
      case 'SHOW_GENOMED':
        return { 'match': { 'genomed': true } }
      case 'SHOW_NOT_GENOMED':
        return { 'match': { 'genomed': false } }
      default:
        return null
    }
  }

  componentWillMount () {
    if (this.state.genes.length || this.state.tags.length) {
      this.fetchArtworks()
    }
  }

  componentDidUpdate (prevProps, prevState) {
    if (
      (this.state.genes !== prevState.genes) ||
      (this.state.tags !== prevState.tags) ||
      (this.state.partner !== prevState.partner) ||
      (this.state.fair !== prevState.fair) ||
      (this.state.publishedFilter !== prevState.publishedFilter) ||
      (this.state.deletedFilter !== prevState.deletedFilter) ||
      (this.state.genomedFilter !== prevState.genomedFilter)
     ) {
      this.fetchArtworks()
    }
  }

  onRemoveGene (geneName) {
    const { genes } = this.state
    this.setState({
      genes: genes.filter(g => g.name !== geneName)
    })
  }

  onAddGene (gene) {
    const { genes } = this.state
    findByName(genes, gene) || this.setState({
      genes: genes.concat(gene)
    })
  }

  onRemoveTag (tagName) {
    const { tags } = this.state
    this.setState({
      tags: tags.filter(t => t.name !== tagName)
    })
  }

  onAddTag (tag) {
    const { tags } = this.state
    findByName(tags, tag) || this.setState({
      tags: tags.concat(tag)
    })
  }

  onSetPartner (partner) {
    this.setState({ partner })
  }

  onClearPartner () {
    this.setState({ partner: null })
  }

  onSetFair (fair) {
    this.setState({ fair })
  }

  onClearFair () {
    this.setState({ fair: null })
  }

  onSetPublishedFilter (filterValue) {
    this.setState({ publishedFilter: filterValue })
  }

  onSetDeletedFilter (filterValue) {
    this.setState({ deletedFilter: filterValue })
  }

  onSetGenomedFilter (filterValue) {
    this.setState({ genomedFilter: filterValue })
  }

  onPreviewArtwork (artwork) {
    this.setState({ previewedArtwork: artwork })
  }

  onPreviewPrevious () {
    const curr = this.state.artworks.indexOf(this.state.previewedArtwork)
    const prev = Math.max(0, curr - 1)
    const artwork = this.state.artworks[prev]
    this.setState({ previewedArtwork: artwork })
  }

  onPreviewNext () {
    const curr = this.state.artworks.indexOf(this.state.previewedArtwork)
    const next = Math.min(this.state.artworks.length, curr + 1)
    const artwork = this.state.artworks[next]
    this.setState({ previewedArtwork: artwork })
  }

  render () {
    const { genes, tags, partner, fair, artworks, previewedArtwork } = this.state
    return (
      <div className='App'>
        <SearchForm
          genes={genes}
          tags={tags}
          partner={partner}
          fair={fair}
          onRemoveGene={this.onRemoveGene}
          onAddGene={this.onAddGene}
          onRemoveTag={this.onRemoveTag}
          onAddTag={this.onAddTag}
          onSetPartner={this.onSetPartner}
          onClearPartner={this.onClearPartner}
          onSetFair={this.onSetFair}
          onClearFair={this.onClearFair}
          publishedFilter={this.state.publishedFilter}
          onSetPublishedFilter={this.onSetPublishedFilter}
          deletedFilter={this.state.deletedFilter}
          onSetDeletedFilter={this.onSetDeletedFilter}
          genomedFilter={this.state.genomedFilter}
          onSetGenomedFilter={this.onSetGenomedFilter}
          />

        <SearchResults
          artworks={artworks}
          previewedArtwork={previewedArtwork}
          onPreviewArtwork={this.onPreviewArtwork}
          onPreviewPrevious={this.onPreviewPrevious}
          onPreviewNext={this.onPreviewNext}
          />
      </div>
    )
  }
}

export default App
