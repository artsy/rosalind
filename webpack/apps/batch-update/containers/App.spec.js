import React from 'react'
import renderer from 'react-test-renderer'
import App from './App'

it('renders correctly', () => {
  const rendered = renderer.create(<App />)
  const tree = rendered.toJSON()
  expect(tree).toMatchSnapshot()
})
