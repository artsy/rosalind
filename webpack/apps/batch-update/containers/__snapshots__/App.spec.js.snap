exports[`test renders correctly 1`] = `
<div
  className="gVLdMb">
  <div
    className="kPSTEV">
    <div
      className="foqgVb">
      <div
        className="PSSoP" />
      <div
        className="GenericAutosuggest">
        <div
          className="react-autosuggest__container">
          <input
            aria-activedescendant={null}
            aria-autocomplete="list"
            aria-expanded={false}
            aria-haspopup={false}
            aria-owns="react-autowhatever-gene-autosuggest"
            autoComplete="off"
            className="react-autosuggest__input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Add a gene"
            role="combobox"
            type="text"
            value="" />
          <div
            className="react-autosuggest__suggestions-container"
            id="react-autowhatever-gene-autosuggest" />
        </div>
      </div>
      <div
        className="GenericAutosuggest">
        <div
          className="react-autosuggest__container">
          <input
            aria-activedescendant={null}
            aria-autocomplete="list"
            aria-expanded={false}
            aria-haspopup={false}
            aria-owns="react-autowhatever-tag-autosuggest"
            autoComplete="off"
            className="react-autosuggest__input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Add a tag"
            role="combobox"
            type="text"
            value="" />
          <div
            className="react-autosuggest__suggestions-container"
            id="react-autowhatever-tag-autosuggest" />
        </div>
      </div>
      <div
        className="GenericAutosuggest">
        <div
          className="react-autosuggest__container">
          <input
            aria-activedescendant={null}
            aria-autocomplete="list"
            aria-expanded={false}
            aria-haspopup={false}
            aria-owns="react-autowhatever-partner-autosuggest"
            autoComplete="off"
            className="react-autosuggest__input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Select a partner"
            role="combobox"
            type="text"
            value="" />
          <div
            className="react-autosuggest__suggestions-container"
            id="react-autowhatever-partner-autosuggest" />
        </div>
      </div>
      <div
        className="GenericAutosuggest">
        <div
          className="react-autosuggest__container">
          <input
            aria-activedescendant={null}
            aria-autocomplete="list"
            aria-expanded={false}
            aria-haspopup={false}
            aria-owns="react-autowhatever-fair-autosuggest"
            autoComplete="off"
            className="react-autosuggest__input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Select a fair"
            role="combobox"
            type="text"
            value="" />
          <div
            className="react-autosuggest__suggestions-container"
            id="react-autowhatever-fair-autosuggest" />
        </div>
      </div>
      <div
        className="iFTuzH">
        <input
          onChange={[Function]}
          onKeyPress={[Function]}
          placeholder="Created after date"
          type="text"
          value="" />
        <div
          className="parsed">
          <a
            href="#"
            onClick={[Function]} />
        </div>
      </div>
      <div
        className="iFTuzH">
        <input
          onChange={[Function]}
          onKeyPress={[Function]}
          placeholder="Created before date"
          type="text"
          value="" />
        <div
          className="parsed">
          <a
            href="#"
            onClick={[Function]} />
        </div>
      </div>
      <div
        className="gxGcSj">
        <div
          className="filter">
          <div>
            Published?
          </div>
          <a
            className="active"
            href="#"
            onClick={[Function]}>
            All
          </a>
          <a
            className={null}
            href="#"
            onClick={[Function]}>
            Published
          </a>
          <a
            className={null}
            href="#"
            onClick={[Function]}>
            Not published
          </a>
        </div>
        <div
          className="filter">
          <div>
            Deleted?
          </div>
          <a
            className="active"
            href="#"
            onClick={[Function]}>
            All
          </a>
          <a
            className={null}
            href="#"
            onClick={[Function]}>
            Deleted
          </a>
          <a
            className={null}
            href="#"
            onClick={[Function]}>
            Not deleted
          </a>
        </div>
        <div
          className="filter">
          <div>
            Genomed?
          </div>
          <a
            className="active"
            href="#"
            onClick={[Function]}>
            All
          </a>
          <a
            className={null}
            href="#"
            onClick={[Function]}>
            Genomed
          </a>
          <a
            className={null}
            href="#"
            onClick={[Function]}>
            Not genomed
          </a>
        </div>
      </div>
      <button
        className="dsgyEb"
        disabled={true}>
        Edit Artworks
      </button>
    </div>
  </div>
  <div
    className="iJVxYF">
    <div
      className="bdWhDJ">
      <div
        className="results" />
    </div>
  </div>
  <div
    className="eGgltq">
    <p>
      0
       works selected
    </p>
    <a
      href="#"
      onClick={[Function]}>
      cancel
    </a>
  </div>
</div>
`;
